File: code/c/test_cribbage.cpp
#include <iostream>
#include <tuple>

#include <cstdlib>
#include <ctime>
#include <cstdlib>

#include "game.h"
#include "policy.h"
#include "greedy.h"
#include "c_policy.h"
#include "c_cribbage.h"

using cpsc474::CribbageGame;
using cpsc474::CompoundPolicy;
using cpsc474::CPolicy;
using cpsc474::GreedyThrower;
using cpsc474::GreedyPegger;
using cpsc474::EvaluationResults;

void reportResults(const EvaluationResults& results);

int main(int argc, char **argv)
{
  size_t numGames = 2;
  
  if (argc > 1)
    {
      numGames = atoi(argv[1]);
    }
  if (numGames <= 0)
    {
      std::cerr << argv[0] << ": number of games must be a positive integer"
		<< std::endl;
      exit(1);
    }
  
  srand(time(NULL));

  CribbageGame game;
  GreedyThrower thrower(game);
  GreedyPegger pegger(game);
  CompoundPolicy benchmark = CompoundPolicy(thrower, pegger);
  CPolicy submission(game, ::keep, ::peg);
  
  EvaluationResults results;

  try
    {
      results = game.evaluatePolicies(&submission, &benchmark, (size_t)numGames);
    }
  catch (const std::string& s)
    {
      std::cerr << s << std::endl;
      exit(1);
    }

  reportResults(results);
}

void reportResults(const EvaluationResults& results)
{
  std::cout << "NET: " << std::get<0>(results).first - std::get<0>(results).second
	    << std::endl
	    << std::get<0>(results).first
	    << "-" << std::get<0>(results).second
	    << " {";
  for (auto i = std::get<1>(results).begin(); i != std::get<1>(results).end(); i++)
    {
      std::cout << (*i).first << "=" << (*i).second << " ";
    }
  std::cout << "} (" << std::get<2>(results) << " hands/game)" << std::endl;
}

File: code/java/TestCribbage.java
import edu.yale.cs.cpsc474.GreedyThrower;
import edu.yale.cs.cpsc474.GreedyPegger;
import edu.yale.cs.cpsc474.CribbagePolicy;
import edu.yale.cs.cpsc474.CompoundPolicy;
import edu.yale.cs.cpsc474.PeggingHistory;
import edu.yale.cs.cpsc474.EvaluationResults;
import edu.yale.cs.cpsc474.CribbageGame;

public class TestCribbage
{
    public static void main(String[] args)
    {
	int numGames = 2;
	int totalGames = 0;
	long runTime = 0;
	if (args.length > 0)
	    {
		if (args[0].equals("--time"))
		    {
			try
			    {
				runTime = Integer.parseInt(args[1]) * 1000L;
				numGames = 1000;
			    }
			catch (NumberFormatException ex)
			    {
				System.err.println("TestCribbage: argument " + args[1] + " is not an integer");
				System.exit(1);
			    }
		    }
		else
		    {
			try
			    {
				numGames = Integer.parseInt(args[0]);
				runTime = 0;
				if (numGames < 0)
				    {
					System.err.println("TestCribbage: number of games must be positive");
				    }
			    }
			catch (NumberFormatException ex)
			    {
				System.err.println("TestCribbage: argument " + args[0] + " is not an integer");
				System.exit(1);
			    }
		    }
	    }

	CribbageGame game = new CribbageGame();
	CribbagePolicy benchmark = new CompoundPolicy(new GreedyThrower(game), new GreedyPegger());
	CribbagePolicy submission = new MyPolicy(game);
	EvaluationResults eval = new EvaluationResults();
	long startTime = System.currentTimeMillis();
	while (totalGames == 0 || System.currentTimeMillis() - startTime < runTime)
	    {
		EvaluationResults results = game.evaluatePolicies(submission, benchmark, numGames);
		eval.add(results);
		totalGames += numGames;
	    }
	System.out.println("NET: " + eval.averageNet());
	System.out.println("CONF: " + (eval.averageNet() - 2 * eval.stddev()));
	System.out.println(eval);
    }
}
